pkg/controller/identity/iamuser/controller.go

```diff
@@ -0,0 +1,251 @@
+	return &external{client: userClient, kube: c.kube}, errors.Wrap(err, errCreateUserClient)
+}
+
+type external struct {
+	kube   client.Client
+	client iam.UserClient
+}
+
+func (e *external) Observe(ctx context.Context, mgd resource.Managed) (managed.ExternalObservation, error) {
+	cr, ok := mgd.(*v1alpha1.User)
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errUnexpectedObject)
+	}
+
+	observed, err := e.client.GetUserRequest(&awsiam.GetUserInput{
+		UserName: aws.String(meta.GetExternalName(cr)),
+	}).Send(ctx)
+
+	if err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(resource.Ignore(iam.IsErrorNotFound, err), errGet)
+	}
+
+	if observed.User == nil {
+		return managed.ExternalObservation{}, errors.New(errSDK)
+	}
+
+	user := *observed.User
+	current := cr.Spec.ForProvider.DeepCopy()
+	iam.LateInitializeUser(&cr.Spec.ForProvider, &user)
+	if !cmp.Equal(current, &cr.Spec.ForProvider) {
+		if err := e.kube.Update(ctx, cr); err != nil {
+			return managed.ExternalObservation{}, errors.Wrap(err, errKubeUpdateFailed)
+		}
+	}
+
+	cr.SetConditions(runtimev1alpha1.Available())
+
+	cr.Status.AtProvider = v1alpha1.UserObservation{
+		Arn:    aws.StringValue(user.Arn),
+		UserID: aws.StringValue(user.UserId),
+	}
+
+	groupsResp, err := e.client.ListGroupsForUserRequest(&awsiam.ListGroupsForUserInput{
+		UserName: aws.String(meta.GetExternalName(cr)),
+	}).Send(ctx)
+
+	if err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(err, errGet)
+	}
+
+	return managed.ExternalObservation{
+		ResourceExists: true,
+		ResourceUpToDate: iam.CompareGroups(cr.Spec.ForProvider, groupsResp.ListGroupsForUserOutput.Groups) &&
+			aws.StringValue(cr.Spec.ForProvider.Path) == aws.StringValue(user.Path),
+	}, nil
+}
+
+func (e *external) Create(ctx context.Context, mgd resource.Managed) (managed.ExternalCreation, error) {
+	cr, ok := mgd.(*v1alpha1.User)
+	if !ok {
+		return managed.ExternalCreation{}, errors.New(errUnexpectedObject)
+	}
+
+	cr.Status.SetConditions(runtimev1alpha1.Creating())
+
+	_, err := e.client.CreateUserRequest(&awsiam.CreateUserInput{
+		Path:                cr.Spec.ForProvider.Path,
+		PermissionsBoundary: cr.Spec.ForProvider.PermissionsBoundary,
+		Tags:                iam.BuildIAMTags(cr.Spec.ForProvider.Tags),
+		UserName:            aws.String(meta.GetExternalName(cr)),
+	}).Send(ctx)
+	return managed.ExternalCreation{}, errors.Wrap(err, errCreate)
+}
+
+func (e *external) Update(ctx context.Context, mgd resource.Managed) (managed.ExternalUpdate, error) {
+	cr, ok := mgd.(*v1alpha1.User)
+	if !ok {
+		return managed.ExternalUpdate{}, errors.New(errUnexpectedObject)
+	}
+
+	groupsResp, err := e.client.ListGroupsForUserRequest(&awsiam.ListGroupsForUserInput{
+		UserName: aws.String(meta.GetExternalName(cr)),
+	}).Send(ctx)
+
+	if err != nil {
+		return managed.ExternalUpdate{}, errors.Wrap(err, errUpdate)
+	}
+
+	if !iam.CompareGroups(cr.Spec.ForProvider, groupsResp.Groups) {
```

Comment:
I wonder whether we should model IAM group membership with a distinct CR, the same way we model role and user policy attachments?

Our rationale with policy attachments was that we get maximum flexibility when the user, policy, and relationship between the two are all distinct CRs. This way you can model only a user, or only a policy in Crossplane, and still create relationships with users or policys that are not modelled with Crossplane. Presumably if we model group membership as an aspect of an IAM user and later added an IAM group CR, we'd have no way to add users to that group unless those users were modelled in Crossplane, right?

Reply:
> Presumably if we model group membership as an aspect of an IAM user and later added an IAM group CR, we'd have no way to add users to that group unless those users were modelled in Crossplane, right?

@negz  I am not sure if I get this completely. I mean if we model the group membership as an aspect of IAMUser now and later add an IAMGroup CR, how would that be a problem?

Reply:
@sahil-lakhwani If we have:

* An `IAMGroup` CR
* An `IAMUser` that may specify the groups the user should be in.

This allows us to:

* Add an `IAMUser` managed by Crossplane to an `IAMGroup` managed by Crossplane.
* Add an `IAMUser` managed by Crossplane to an IAM group not managed by Crossplane.

However it does not allow us to add an IAM user that is not managed by Crossplane to an `IAMGroup` that is managed by Crossplane; the `IAMUser` must be modelled by Crossplane.

If we instead had:

* An `IAMGroup` CR.
* An `IAMUser` CR`.
* An `IAMGroupMembership` CR.

We would be able to:

* Add an `IAMUser` managed by Crossplane to an `IAMGroup` managed by Crossplane.
* Add an `IAMUser` managed by Crossplane to an IAM group not managed by Crossplane.
* Add an IAM user not managed by Crossplane to an `IAMGroup` managed by Crossplane.
* Add an IAM user not managed by Crossplane to an IAM group not managed by Crossplane.

Reply:
Thanks @negz . I have removed the group membership for now.